Proposed documentation for proposed Model element

zachernuk opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

This PR adds API documentation about the proposed HTML Model element, a browser-native feature for the in-line display of 3D content. It includes illustrations of some core concepts like perspective, lighting and appropriate file formats. 

<!-- ✍️ Summarize your changes in one or two sentences -->

### Motivation

The Model element is a totally new web feature and it'll be exceptionally difficult to add documentation in a piecemeal fashion. I'm hoping that this covers the core concepts enough to author new pages from, and matches the MDN house style well _enough_ that other folks can improve them to the point that they are real!
<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

I am submitting provisional documentation for the HTML model element, a feature that is the major source of discussion in the W3C Immersive Web CG (https://github.com/immersive-web/model-element/ )
and which is available as a feature preview in visionOS 2.4. I am an editor of the specification and have been facilitating the discussion around it.

While the specification doesn’t have a lot of detail, the API in visionOS is based on the consensus discussion in the CG, and is described in the Model Element explainer: (https://github.com/immersive-web/model-element/blob/main/explainer.md )

There is also a WebGL/WebXR-based interactive explainer demo available here:
https://immersive-web.github.io/model-element/explainer_demo.html

We also have public samples being contributed by multiple browser vendors in the model element samples repository.
https://immersive-web.github.io/model-element-samples/ 

We have previously submitted related additions to the WHATWG HTML specification as via an introductory issue and initial PR for the specification stub.
https://github.com/whatwg/html/issues/10901
https://github.com/whatwg/html/pull/11191
 


### Related issues and pull requests

#### WHATWG/HTML
* https://github.com/whatwg/html/issues/10901
* https://github.com/whatwg/html/pull/11191

#### the Web Platform Tests repository
* https://github.com/web-platform-tests/wpt/pull/51667

#### HTML-ARIA
* https://github.com/w3c/aria/issues/2508

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

OnkarRuikar committed:
Synchronize with BCD v6.0.17 (#39655)

Co-authored-by: OnkarRuikar <OnkarRuikar@users.noreply.github.com>

Rageking8 committed:
Fix typos (#39658)

* Fix 2 typos

* Fix more

* Fix more

---------

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

Psychpsyo committed:
Fix determiner choice with HTML elements (#39660)

* Fix an typo

* Fix more

---------

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

mdn-bot committed:
Update InterfaceData based on WebRef (#39654)

hamishwillee committed:
FF139 SVG discard element and api removed - expr feature too (#39636)

* Remove discard from experimental features

* Delete <discard> SVG element

* Remove or unlink references to <discard> element

* SVGDiscardElement - mark deprecated and remove links to discard element

Josh-Cena committed:
Remove link to nonexistent page (#39656)

Fix content issues

hamishwillee committed:
FF140 Relnote: Notification.maxActions unpublished to nightly (#39640)

* FF140 Relnote: Notification.maxActions unpublished to nightly

* Update files/en-us/mozilla/firefox/releases/140/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

* Update actions and maxActions in Notifications

---------

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

samuel871211 committed:
doc[sandbox]: remove `HTMLDialogElement.showModal` from `allow-popups` (#39586)

* doc[sandbox]: remove `HTMLDialogElement.showModal` from `allow-popups`

* tidying layout. Ignore.

* Update files/en-us/web/http/reference/headers/content-security-policy/sandbox/index.md

---------

Co-authored-by: Hamish Willee <hamishwillee@gmail.com>

Josh-Cena committed:
Make String.p.pad{End,Start} articles consistent (#39677)

chrisdavidmills committed:
Improve version control consistentency (#39616)

* Improve version control consistentency

* Update files/en-us/learn_web_development/core/version_control/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

* Update files/en-us/learn_web_development/core/version_control/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

* Update files/en-us/learn_web_development/core/version_control/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

* Update files/en-us/learn_web_development/core/version_control/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

* Update files/en-us/learn_web_development/core/version_control/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

* Update files/en-us/learn_web_development/core/version_control/index.md

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

* Last little bit of restructuring

---------

Co-authored-by: Dipika Bhattacharya <dipika@foss-community.org>

HomyeeKing committed:
Update index.md (#39684)

* Update index.md

revert https://github.com/mdn/content/pull/39573

* Fix linting error

---------

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

dependabot[bot] committed:
Bump codelytv/pr-size-labeler from 1.10.2 to 1.10.3 (#39686)

dipikabh committed:
Add the default value indicator in prose (#39626)

* add default value indicator

* Apply suggestions from code review

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

---------

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

Ma-ve committed:
fix: typo in options_ui deprecation warning (#39681)

hamishwillee committed:
FF140 Relnote - TaskSignal.any() (#39635)

hamishwillee committed:
FF140 Relnote: Serializes < and > in attributes (#39638)

hamishwillee committed:
FF140 TaskSignal.any() - added to expr feat (#39634)

hamishwillee committed:
FF140 escaping `<` and `>` to `&lt;` and `&gt;` in attributes when serializing HTML (#39639)

K0R0BKIN committed:
Fix typo (#39691)

hamishwillee committed:
FF140 Relnote/Expr Atomics.waitAsync() (#39678)

Co-authored-by: Vadim Makeev <hi@pepelsbey.dev>

hamishwillee committed:
FF140 Relnote: Cookie Store API (#39676)

* FF140 Relnote: Cookie Store API

* Tidy top level doc

* Update index.md

---------

Co-authored-by: Vadim Makeev <hi@pepelsbey.dev>

mohammad-kenaan committed:
Add a missing backtick at the end of a code element (#39693)

fix: add missing backtick at the end of code element

cssinate committed:
Add view-transition-class property (#37901)

* Add view-transition-class property

* Address comments from @estelle

* Fixing awkward wording

* Add backticks to period

* Remove KeyboardEvent.initKeyEvent page (#39027)

* Adding suggested changes

* fix linting issues

* Apply suggestions from code review

Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>

* Apply suggestions from code review

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

* chore(css): Use parens for view transition pseudo names

* Update files/en-us/web/api/view_transition_api/index.md

* chore(css): ad param section

* chore(css): try to make the view transition names more illustrative

* Update files/en-us/web/css/view-transition-class/index.md

---------

Co-authored-by: Paul Grant <paul.grant@hootsuite.com>
Co-authored-by: Joshua Chen <sidachen2003@gmail.com>
Co-authored-by: Estelle Weyl <estelle@weyl.org>
Co-authored-by: Brian Smith <brian@smith.berlin>
Co-authored-by: Estelle Weyl <estelle@openwebdocs.org>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump @octokit/rest from 21.1.1 to 22.0.0 (#39687)

Bumps [@octokit/rest](https://github.com/octokit/rest.js) from 21.1.1 to 22.0.0.
- [Release notes](https://github.com/octokit/rest.js/releases)
- [Commits](https://github.com/octokit/rest.js/compare/v21.1.1...v22.0.0)

---
updated-dependencies:
- dependency-name: "@octokit/rest"
  dependency-version: 22.0.0
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

eyad-alkhalidy committed:
Fix typo: Remove a duplicated word (#39699)

fix(typo): remove a duplicated word

Josh-Cena committed:
Array.fromAsync() mapFn does not always have its input awaited (#39690)

* Array.fromAsync() mapFn does not always have its input awaited

* Update files/en-us/web/javascript/reference/global_objects/array/fromasync/index.md

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

---------

Co-authored-by: Chris Mills <chrisdavidmills@gmail.com>

Josh-Cena committed:
Remove unused assets in <discard> folder (#39696)

Chevindu committed:
Replace "blue" with "red" where referring to the hex code #a71e14 (#39700)

Replace blue with red where referring to the hex code #a71e14

hamishwillee committed:
FF140 Relnote/Expr Feat: Dialog closeBy attribute (#39680)

* FF140 Relnote/Expr Feat: Dialog closeBy attribute

* Update relnote too

* Update files/en-us/mozilla/firefox/experimental_features/index.md

* Update index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Vadim Makeev <hi@pepelsbey.dev>

---------

Co-authored-by: Vadim Makeev <hi@pepelsbey.dev>
Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

hamishwillee committed:
Atomics.wait()/waitAsync() improvements (#39679)

* Atomics.wait()/waitAsync() improvements

* Update files/en-us/web/javascript/reference/global_objects/atomics/wait/index.md

* Update files/en-us/web/javascript/reference/global_objects/atomics/wait/index.md

* Remove the note

* waitAsync returns and object

* Update files/en-us/web/javascript/reference/global_objects/atomics/wait/index.md

* Apply suggestions from code review

* Update index.md

* Update index.md

* Update index.md

---------

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

estelle committed:
Add years to external article links (#39701)

* Add years to external article links

* typo

* Update files/en-us/web/css/font-smooth/index.md

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

* Update files/en-us/web/css/font-smooth/index.md

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

---------

Co-authored-by: Joshua Chen <sidachen2003@gmail.com>

Josh-Cena committed:
Lint: prefer for...of loops over plain for loops (#39661)

Josh-Cena committed:
Lint: use strict equality, avoid numeric loss of precision (#39667)

* Lint: use strict equality, avoid numeric loss of precision

* Fix

Josh-Cena committed:
Lint: use modern syntax, object shorthand/exponentiation/regex literal (#39657)

Josh-Cena committed:
Lint: prefer Object.hasOwn, avoid eval (#39668)

Josh-Cena committed:
Move punctuation out of code spans, fix bad backticks (#39704)

hamishwillee committed:
FF139 Relnote - SVG discard removed (#39637)

* FF139 Relnote - SVG discard removed

* Update files/en-us/mozilla/firefox/releases/139/index.md

Co-authored-by: Brian Smith <brian@smith.berlin>

* Update files/en-us/mozilla/firefox/releases/139/index.md

Co-authored-by: Brian Smith <brian@smith.berlin>

* Update files/en-us/mozilla/firefox/releases/139/index.md

---------

Co-authored-by: Brian Smith <brian@smith.berlin>

zachernuk committed:
Initial suite of files to describe the Model element

A full suite of the currently agreed-upon API and capability of the Model element, as discussed in the Immersive Web CG

zachernuk committed:
Update yarn.lock

rebuilt with the yarn builder so the lockfile should match upstream

zachernuk committed:
Update yarn.lock

Copied the yarn lockfile back to the upstream state

Josh-Cena committed:
Lint: fix CSS syntax and some antipatterns (#39685)

bsmth committed:
chore(fx): pages missing text before first h2 (#39712)

zachernuk committed:
Fix from GH-actions feedback

Updated casing on some attribute/property names, update to the richer paths on things like Content categories

zachernuk committed:
Merge branch 'main' into model-element

Josh-Cena commented on yarn.lock:
```diff
@@ -9836,4 +9836,4 @@ yoctocolors@^2.0.0:
   version "2.0.4"
   resolved "https://registry.yarnpkg.com/zwitch/-/zwitch-2.0.4.tgz#c827d4b0acb76fc3e685a4c6ec2902d51070e9d7"
-  integrity sha512-bXE4cR/kVZhKZX/RjPEflHaKVhUVl85noU3v6b8apfQEc1x4A+zBxjZ4lN8LqGd6WZ3dl98pY4o717VFmoPp+A==
+  integrity sha512-bXE4cR/kVZhKZX/RjPEflHaKVhUVl85noU3v6b8apfQEc1x4A+zBxjZ4lN8LqGd6WZ3dl98pY4o717VFmoPp+A==
```
Revert

Josh-Cena commented on files/en-us/web/html/reference/elements/model/index.md:
```diff
@@ -0,0 +1,279 @@
+
+- While it provides benefits for all devices and User Agents, the `<model>` element has particular value on platforms with stereoscopic and/or head-tracking capabilities, such as Head-Mounted Displays (HMDs). It can be used as declarative alternative to the WebXR API to provide spatial content.
+
+## Object fit
```
These should be H3 headings. The H2 sequence is exactly those defined in the page template

Josh-Cena commented on files/en-us/web/api/htmlmodelelement/width/index.md:
```diff
@@ -0,0 +1,62 @@
+
+### Initial presentation
+
+A model's contents is set to automatically fit within the portal's
+`width` and `height` dimensions at the moment that the `ready` Promise
```
General comment: please don't wrap your prose like this. Either wrap at each period, or don't wrap and just have paragraph breaks

Josh-Cena commented on files/en-us/web/api/htmlmodelelement/stagemode/index.md:
```diff
@@ -0,0 +1,62 @@
+
+Setting the `stagemode` to `orbit` immediately has these consequences for the model element's `entityTransform`:
+
+- Transform is <strong>`orbit-fit`</strong>:
```
Use Markdown

Josh-Cena commented on files/en-us/web/api/htmlmodelelement/ready/index.md:
````diff
@@ -0,0 +1,48 @@
+This example demonstrates how to confirm that both a model's scene and environment map resources are ready for rendering, and how to gracefully address failed parsing:
+
+```js
+let modelElem = document.getElementById("model");
````
Use `const` where possible

Josh-Cena commented on files/en-us/web/api/htmlmodelelement/environmentmapready/index.md:
```diff
@@ -0,0 +1,54 @@
+
+## See also
+
+- [model.ready](/en-US/docs/Web/API/HTMLModelElement/ready)
```
```suggestion
- [`HTMLModelElement.ready`](/en-US/docs/Web/API/HTMLModelElement/ready)
```

Josh-Cena commented on files/en-us/web/api/htmlmodelelement/entitytransform/index.md:
```diff
@@ -0,0 +1,65 @@
+
+## See also
+
+- {{domxref("HTMLModelElement")}}: Interface used to define the `HTMLModelElement.entityTransform` property
```
```suggestion
- {{domxref("HTMLModelElement")}}
```

We don't add explanatory text for see also

Josh-Cena reviewed (commented):
Some random drive-by reviews. Not comprehensive in any way, but they indicate some common problems that would need to be fixed throughout.

Also we need browser compat data for this. It should be shipping in at least one browser to be documented.

zachernuk committed:
Update with @Josh-Cena 's feedback

Remove arbitrary wrapping,
Change `let` to `const`,
Fixed the 'ARIA role' reference urls,
Demoted some headings to be in line with MDN standards,
Applied markdown-based bolding

zachernuk committed:
Merge branch 'model-element' of https://github.com/zachernuk/content into model-element

zachernuk committed:
Merge branch 'model-element' of https://github.com/zachernuk/content into model-element

zachernuk committed:
Merge branch 'mdn:main' into model-element

zachernuk commented:
I wasn't able to work out how to revert the single file within the overall PR. I'm going to close this and move the work into this PR: https://github.com/mdn/content/pull/39811
